src/TranslationDispatcher.ts

```diff
@@ -0,0 +1,94 @@
+import { DOMNodesTranslator } from './DOMNodesTranslator';
+import { IntersectionObserverWithFilter } from './IntersectionObserverWithFilter';
+import { isIntersectableNode } from './utils/isIntersectableNode';
+import { visitWholeTree } from './utils/visitWholeTree';
+
+export type TranslatableNodePredicate = (node: Node) => boolean;
+
+/**
+ * Coordinates the DOM nodes translation process.
+ * Chooses between lazy and immediate translation; defaults to immediate if not provided.
+ */
+export class TranslationDispatcher {
+	private readonly filter;
+	private readonly nodeTranslator;
+	// if dependency is not passed, then the node will not be translated lazy
+	private readonly lazyTranslator;
+
+	constructor({
+		filter,
+		nodeTranslator,
+		lazyTranslator,
+	}: {
+		filter: TranslatableNodePredicate;
+		nodeTranslator: DOMNodesTranslator;
+		lazyTranslator?: IntersectionObserverWithFilter;
+	}) {
+		this.filter = filter;
+		this.nodeTranslator = nodeTranslator;
+		this.lazyTranslator = lazyTranslator || null;
+	}
+
+	public updateNode(node: Node) {
+		this.nodeTranslator.updateNode(node);
+	}
+
+	public hasNode(node: Node) {
+		return this.nodeTranslator.hasNode(node);
+	}
+	/**
+	 * Translates the node and all its nested translatable nodes (text and attribute nodes)
+	 */
+	public translateNode(node: Node) {
+		// Translate all nodes which element contains (text nodes and attributes of current and inner elements)
+		if (node instanceof Element) {
+			visitWholeTree(node, (node) => {
+				if (node instanceof Element) return;
+				this.translateNode(node);
+			});
+			return;
+		}
+		// Handle text nodes and attributes
+
+		// Skip node if it does not satisfy the filter
+		if (!this.filter(node)) return;
+		if (this.lazyTranslator) {
+			// Lazy translate when own element intersect viewport
+			// But translate at once if node have not parent (virtual node) or parent node is outside of body (utility tags like meta or title)
+			const isAttachedToDOM = node.getRootNode() !== node;
+			const observableNode =
+				node instanceof Attr ? node.ownerElement : node.parentElement;
+
+			// Ignore lazy translation for non-intersecting nodes and translate it immediately
+			if (
+				isAttachedToDOM &&
+				observableNode !== null &&
+				isIntersectableNode(observableNode)
```

Comment:
Move this logic to `NodeIntersectionObserver` and in case node have no owner or it is not intersectable, callback must be called immediately